src/core/border_router/routing_manager.cpp

```diff
@@ -0,0 +1,739 @@
+        }
+        else
+        {
+            otLogCritBr("failed to send %uth Router Solicitation: %s", mRouterSolicitCount,
+                        otThreadErrorToString(error));
+        }
+
+        nextSolicitationDelay =
+            (mRouterSolicitCount == kMaxRtrSolicitations) ? kMaxRtrSolicitationDelay : kRtrSolicitationInterval;
+
+        otLogDebgBr("Router Solicitation timer scheduled in %u seconds", nextSolicitationDelay);
+        mRouterSolicitTimer.Start(nextSolicitationDelay * 1000);
+    }
+    else
+    {
+        // Re-evaluate our routing policy and send Router Advertisement if necessary.
+        EvaluateRoutingPolicy();
+    }
+}
+
+void RoutingManager::HandleDiscoveredOnLinkPrefixInvalidTimer(Timer &aTimer)
+{
+    aTimer.GetOwner<RoutingManager>().HandleDiscoveredOnLinkPrefixInvalidTimer();
+}
+
+void RoutingManager::HandleDiscoveredOnLinkPrefixInvalidTimer(void)
+{
+    mDiscoveredOnLinkPrefix.Clear();
+
+    // The discovered on-link prefix becomes invalid, start Router Solicitation
+    // to discover new one.
+    StartRouterSolicitation();
+}
+
+void RoutingManager::HandleRouterSolicit(const Ip6::Address &aSrcAddress,
+                                         const uint8_t *     aBuffer,
+                                         uint16_t            aBufferLength)
+{
+    OT_UNUSED_VARIABLE(aSrcAddress);
+    OT_UNUSED_VARIABLE(aBuffer);
+    OT_UNUSED_VARIABLE(aBufferLength);
+
+    otLogInfoBr("received Router Solicitation from %s on interface %s", aSrcAddress.ToString().AsCString(),
+                GetInfraIfName());
+
+    mRouterAdvertisementTimer.Start(Random::NonCrypto::GetUint32InRange(0, kMaxRaDelayTime));
+}
+
+bool RoutingManager::HandlePrefixInfoOption(const RouterAdv::PrefixInfoOption &aPio)
+{
+    bool      needReevaluate = false;
+    TimeMilli expireTime;
+
+    if (aPio.GetValidLifetime() == 0)
+    {
+        if (mDiscoveredOnLinkPrefix.IsEqual(aPio.GetPrefix(), aPio.GetPrefixLength()))
+        {
+            otLogInfoBr("invalidate discovered on-link prefix: %s", mDiscoveredOnLinkPrefix.ToString().AsCString());
+            mDiscoveredOnLinkPrefix.Clear();
+            mDiscoveredOnLinkPrefixInvalidTimer.Stop();
+            StartRouterSolicitation();
+        }
+        ExitNow();
+    }
+
+    expireTime = GetPrefixExpireTime(aPio.GetValidLifetime());
+
+    if (!IsValidOnLinkPrefix(mDiscoveredOnLinkPrefix) ||
+        (mDiscoveredOnLinkPrefixInvalidTimer.IsRunning() &&
+         expireTime > mDiscoveredOnLinkPrefixInvalidTimer.GetFireTime()))
+    {
+        mDiscoveredOnLinkPrefix.Set(aPio.GetPrefix(), aPio.GetPrefixLength());
+
+        otLogInfoBr("set discovered on-link prefix to %s, valid lifetime: %u seconds",
+                    mDiscoveredOnLinkPrefix.ToString().AsCString(), aPio.GetValidLifetime());
+
+        if (aPio.GetValidLifetime() == RouterAdv::kInfiniteLifetime)
+        {
+            mDiscoveredOnLinkPrefixInvalidTimer.Stop();
```

Comment:
Once we receive a PIO with `kInfiniteLifetime`, we are going to stop `mDiscoveredOnLinkPrefixInvalidTimer`. In such case, we are never going to invalidate `mDiscoveredOnLinkPrefix`. If the router that was advertising `mDiscoveredOnLinkPrefix` is shutdown without having the chance to send the invalidating PIO, the BR will end up not advertising `On Link Prefix` forever. 

Should we address such case?

A possible solution might be sending RS timely to make sure that the `On Link Prefix` is still being advertised by another router.

Reply:
The on-link prefix can continue working event if the router advertised it has gone. The problem is that, new BRs joined the WiFi link will not be able to find the on-link prefix and thus a new on-link prefix will be advertised by the new BR. So that we can get our discovered on-link prefix updated.